src/rust/engine/src/nodes.rs

```diff
@@ -1026,37 +1070,67 @@ impl WrappedNode for Task {
     let product = self.product;
     let can_modify_workunit = self.task.can_modify_workunit;
 
-    let result_val =
-      externs::call_function(&externs::val_for(&func.0), &deps).map_err(Failure::from_py_err)?;
-    let mut result_val: Value = result_val.into();
-    let mut result_type = externs::get_type_for(&result_val);
-    if result_type == context.core.types.coroutine {
-      result_val = Self::generate(context.clone(), params, entry, result_val).await?;
-      result_type = externs::get_type_for(&result_val);
-    }
+    let result_val = {
+      // All of the params should have been supplied by coroutines. Ensure that all of the
+      // coroutines succeeded, otherwise error.
+      let mut param_deps: Vec<Value> = vec![];
+      for result in deps.into_iter() {
+        match result {
+          InvocationResult::Exception(err) => {
+            return Ok(PythonRuleOutput::create_rule_output(
+              InvocationResult::Exception(err),
+              &context.core,
+              can_modify_workunit,
+            ));
+          }
+          InvocationResult::SuccessfulReturn(val) => {
+            param_deps.push(val);
+          }
+        }
+      }
+      // The params were all successfully collected, now call the current @rule function with them.
+      externs::call_catching_error(&externs::val_for(&func.0), &param_deps)
```

Comment:
Since `call_catching_error` is just applying an `.into()` to convert the `call_function` `Result` type into an `InvocationResult`, and since we only use it in this one place, I think it's cleaner to avoid defining the function and just handle the type-conversion from `call_function` here. In general I've been trying to make the functions we have in `externs` that deal with python objects fewer and more general, and letting each call site manipulate the types to whatever is required at that call site.

Reply:
I have done this, along with further cleanup of the code that executes `externs::call_function()`.